core/src/main/scala/kafka/log/remote/RemoteIndexCache.scala

```diff
@@ -0,0 +1,284 @@
+            index.get.sanityCheck()
+            index
+          }
+
+          val timeIndex: LazyIndex[TimeIndex] = {
+            val index = LazyIndex.forTime(timestampIndexFile, offset, Int.MaxValue, writable = false)
+            index.get.sanityCheck()
+            index
+          }
+
+          val txnIndex: TransactionIndex = {
+            val index = new TransactionIndex(offset, txnIndexFile)
+            index.sanityCheck()
+            index
+          }
+
+          val entry = new Entry(offsetIndex, timeIndex, txnIndex)
+          entries.put(uuid, entry)
+        } else {
+          // Delete all of them if any one of those indexes is not available for a specific segment id
+          Files.deleteIfExists(offsetIndexFile.toPath)
+          Files.deleteIfExists(timestampIndexFile.toPath)
+          Files.deleteIfExists(txnIndexFile.toPath)
+        }
+      }
+    })
+  }
+
+  init()
+
+  // Start cleaner thread that will clean the expired entries
+  val cleanerThread: ShutdownableThread = new ShutdownableThread("remote-log-index-cleaner") {
+    setDaemon(true)
+
+    override def doWork(): Unit = {
+      while (!closed) {
+        try {
+          val entry = expiredIndexes.take()
+          info(s"Cleaning up index entry $entry")
+          entry.cleanup()
+        } catch {
+          case ex: InterruptedException => info("Cleaner thread was interrupted", ex)
+          case ex: Exception => error("Error occurred while fetching/cleaning up expired entry", ex)
+        }
+      }
+    }
+  }
+  cleanerThread.start()
+
+  def getIndexEntry(remoteLogSegmentMetadata: RemoteLogSegmentMetadata): Entry = {
+    def loadIndexFile[T](fileName: String,
+                         suffix: String,
+                         fetchRemoteIndex: RemoteLogSegmentMetadata => InputStream,
+                         readIndex: File => T): T = {
+      val indexFile = new File(cacheDir, fileName + suffix)
+
+      def fetchAndCreateIndex(): T = {
+        val tmpIndexFile = new File(cacheDir, fileName + suffix + RemoteIndexCache.TmpFileSuffix)
+
+        val inputStream = fetchRemoteIndex(remoteLogSegmentMetadata)
+        try {
+          Files.copy(inputStream, tmpIndexFile.toPath)
+        } finally {
+          if (inputStream != null) {
+            inputStream.close()
+          }
+        }
+
+        Utils.atomicMoveWithFallback(tmpIndexFile.toPath, indexFile.toPath, false)
+        readIndex(indexFile)
+      }
+
+      if (indexFile.exists()) {
+        try {
+          readIndex(indexFile)
+        } catch {
+          case ex: CorruptRecordException =>
+            info("Error occurred while loading the stored index", ex)
+            fetchAndCreateIndex()
+        }
+      } else {
+        fetchAndCreateIndex()
+      }
+    }
+
+    lock synchronized {
+      entries.computeIfAbsent(remoteLogSegmentMetadata.remoteLogSegmentId().id(), (uuid: Uuid) => {
```

Comment:
Thanks for the update, but I think we still lock for the period of loading 3 index files. How about this:
```java
if (!entries.containsKey(remoteLogSegmentMetadata.remoteLogSegmentId().id())) {
        val startOffset = remoteLogSegmentMetadata.startOffset()
        // uuid.toString uses URL encoding which is safe for filenames and URLs.
        val fileName = startOffset.toString + "_" + uuid.toString + "_"

        val offsetIndex: LazyIndex[OffsetIndex] = loadIndexFile(fileName, UnifiedLog.IndexFileSuffix,
          rlsMetadata => remoteStorageManager.fetchIndex(rlsMetadata, IndexType.OFFSET),
          file => {
            val index = LazyIndex.forOffset(file, startOffset, Int.MaxValue, writable = false)
            index.get.sanityCheck()
            index
          })

        val timeIndex: LazyIndex[TimeIndex] = loadIndexFile(fileName, UnifiedLog.TimeIndexFileSuffix,
          rlsMetadata => remoteStorageManager.fetchIndex(rlsMetadata, IndexType.TIMESTAMP),
          file => {
            val index = LazyIndex.forTime(file, startOffset, Int.MaxValue, writable = false)
            index.get.sanityCheck()
            index
          })

        val txnIndex: TransactionIndex = loadIndexFile(fileName, UnifiedLog.TxnIndexFileSuffix,
          rlsMetadata => remoteStorageManager.fetchIndex(rlsMetadata, IndexType.TRANSACTION),
          file => {
            val index = new TransactionIndex(startOffset, file)
            index.sanityCheck()
            index
          })

          // start lock here
          lock synchronized {
               if (!entries.containsKey(remoteLogSegmentMetadata.remoteLogSegmentId().id())) {
                    entries.put(remoteLogSegmentMetadata.remoteLogSegmentId().id(), new Entry(offsetIndex, timeIndex, txnIndex));
               }
          }
}
```

WDYT?

Reply:
This creates race between writers and readers of these files which may fail readers with IO errors. 

For ex: The race can occur in 3 ways
1. Multiple writers writing those indexes concurrently
2. One of them will update the entry with the existing file 
   But that file could have been over written in step-1
3. When readers start reading, those file might have been overwritten in step-1.

Reply:
https://issues.apache.org/jira/browse/KAFKA-13560 covers to get indexes or data in async manner for critical paths, we can cover as part of that JIRA. wdyt?

Reply:
Sounds good to me! I'll resolve this comment.